src/libxrpl/protocol/STEitherAmount.cpp

```diff
@@ -0,0 +1,425 @@
+    {
+        std::string val = v.asString();
+        std::vector<std::string> elements;
+        boost::split(elements, val, boost::is_any_of("\t\n\r ,/"));
+
+        if (elements.size() > 3)
+            Throw<std::runtime_error>("invalid amount string");
+
+        value = elements[0];
+
+        if (elements.size() > 1)
+            currencyOrMPTID = elements[1];
+
+        if (elements.size() > 2)
+            issuer = elements[2];
+    }
+    else
+    {
+        value = v;
+    }
+
+    bool const native = !currencyOrMPTID.isString() ||
+        currencyOrMPTID.asString().empty() ||
+        (currencyOrMPTID.asString() == systemCurrencyCode());
+
+    if (native)
+    {
+        if (v.isObjectOrNull())
+            Throw<std::runtime_error>("XRP may not be specified as an object");
+        issue = xrpIssue();
+    }
+    else
+    {
+        if (isMPT)
+        {
+            // sequence (32 bits) + account (160 bits)
+            MPTID u;
+            if (!u.parseHex(currencyOrMPTID.asString()))
+                Throw<std::runtime_error>("invalid MPTokenIssuanceID");
+            issue = u;
+        }
+        else
+        {
+            issue = Issue{};
+            if (!to_currency(
+                    std::get<Issue>(issue).currency,
+                    currencyOrMPTID.asString()))
+                Throw<std::runtime_error>("invalid currency");
+            if (!issuer.isString() ||
+                !to_issuer(std::get<Issue>(issue).account, issuer.asString()))
+                Throw<std::runtime_error>("invalid issuer");
+            if (isXRP(std::get<Issue>(issue)))
+                Throw<std::runtime_error>("invalid issuer");
+        }
+    }
+
+    if (value.isInt())
+    {
+        if (value.asInt() >= 0)
+        {
+            mantissa = value.asInt();
+        }
+        else
+        {
+            mantissa = -value.asInt();
+            negative = true;
+        }
+    }
+    else if (value.isUInt())
+    {
+        mantissa = v.asUInt();
+    }
+    else if (value.isString())
+    {
+        if (std::holds_alternative<Issue>(issue))
+        {
+            STAmount const ret =
+                amountFromString(std::get<Issue>(issue), value.asString());
+            mantissa = ret.mantissa();
+            exponent = ret.exponent();
+            negative = ret.negative();
+        }
+        else
+        {
+            STMPTAmount const ret =
+                amountFromString(std::get<MPTIssue>(issue), value.asString());
+            negative = ret.value() < 0;
+            mantissa = !negative ? ret.value() : -ret.value();
+            exponent = 0;
+        }
+    }
+    else
+    {
+        Throw<std::runtime_error>("invalid amount type");
+    }
+
+    if (std::holds_alternative<Issue>(issue))
+        return STEitherAmount{
+            name,
+            STAmount{
+                std::get<Issue>(issue), mantissa, exponent, native, negative}};
+    while (exponent-- > 0)
+        mantissa *= 10;
```

Comment:
What if `mantissa  > maxMPTokenAmount` prior to entering this loop, or becomes so after the `*= 10` but before `exponent == 0`?

Unless there is other logic to prevent this I think we need to guard against `mantissa *= 10` overflowing the `uint64_t`.  This is only guaranteed if `mantissa <= maxMPTokenAmount` prior to the `*= 10`.